packages/esm/core/src/Engines/Extensions/query/query.base.ts

```diff
@@ -0,0 +1,363 @@
+import type { Nullable, int } from "@babylonjs/core/types.js";
+import type { Observer } from "@babylonjs/core/Misc/observable.js";
+import { PerfCounter } from "@babylonjs/core/Misc/perfCounter.js";
+import type { _TimeToken } from "@babylonjs/core/Instrumentation/timeToken.js";
+import type { IBaseEnginePublic } from "../../engine.base.js";
+import { AbstractMesh } from "@babylonjs/core/Meshes/abstractMesh.js";
+
+// TODO there is a breaking change here - returning engine state instead of engine

```

Comment:
Yes and no. is ESM there is no issue with breaking changes. BUT when converting it back to UMD there might be a need for a proxy between the type that is returned and a ThinEngine, for example. One way of solving this might be to use ReturnType or some other Tyepscript helper, but I still have a few things to solve before fixing this.

The same issue can be seen in any observable that notifies an Engine. in ESM it will notify the state object but in UMD it will notify with a fully constructed engine. Though their types might be related, it is a totally different typescript type.

I hope it's clear :-) . I will be happy to explain it better